// node-omxplayer/index.js
'use strict';

// ----- Requires ----- //

let spawn = require('child_process').spawn;
let EventEmitter = require('events');
let StringDecoder = require('string_decoder').StringDecoder

// let decoder = new StringDecoder('utf-8')


// ----- Setup ----- //

// The permitted audio outputs, local means via the 3.5mm jack.
let ALLOWED_OUTPUTS = ['hdmi', 'local', 'both', 'alsa'];


// ----- Functions ----- //

// Creates an array of arguments to pass to omxplayer.
function buildArgs (source, givenOutput, loop, initialVolume, showOsd) {
	let output = '';

	if (givenOutput) {

		if (ALLOWED_OUTPUTS.indexOf(givenOutput) === -1 && ! givenOutput.match(/alsa:.*/)) {
			throw new Error(`Output ${givenOutput} not allowed.`);
		}

		output = givenOutput;

	} else {
		output = 'local';
	}

	let osd = false;
	if (showOsd) {
		osd = showOsd;
	}

	let args = [source, '-o', output, '--blank', osd ? '' : '--no-osd'];

	// Handle the loop argument, if provided
	if (loop) {
		args.push('--loop');
	}

	// Handle the initial volume argument, if provided
	if (Number.isInteger(initialVolume)) {
		args.push('--vol', initialVolume);
	}

	return args;

}


// ----- Omx Class ----- //

function Omx (source, output, loop, initialVolume, showOsd) {

	// ----- Local Vars ----- //

	let player = null;
	let omxplayer = new EventEmitter();
	let open = false;
	let volume = initialVolume;
  omxplayer["open"] = open;
	omxplayer["pid"] = false

	// ----- Local Functions ----- //

	// Marks player as closed.
	function updateStatus () {

		open = false;
		omxplayer["open"] = open;
		omxplayer.emit('close');

	}

	// Emits an error event, with a given message.
	function emitError (message) {

		open = false;
		omxplayer["open"] = open;
		omxplayer.emit('error', message);

	}

	function emitPlaying () {

		open = true;
		omxplayer["open"] = open;
		omxplayer.emit('playback');

	}

	// Spawns the omxplayer process.
	function spawnPlayer (src, out, loop, initialVolume, showOsd) {

		let build_args = buildArgs(src, out, loop, initialVolume, showOsd);

		let stdbuf_args = new Array(
			'-i0', '-o0', '-e0',
			'/usr/bin/omxplayer.bin'
		)

		let args = stdbuf_args.concat(build_args)

		console.log('args for omxplayer:', args);
		let omxProcess = spawn('/usr/bin/stdbuf', args, {detached: true});
		omxplayer["pid"] = omxProcess["pid"]

		open = true;
		omxplayer["open"] = open;
		emitPlaying()
		//emit player open/playing

		omxProcess.stdin.setEncoding('utf-8');
		omxProcess.on('close', updateStatus);


		omxProcess.on('error', () => {
			emitError('Problem running omxplayer, is it installed?.');
		});

		omxplayer["process"] = omxProcess

		return omxProcess;

	}

	// Simulates keypress to provide control.
	var queryRunning = false;
  var stdQuery = new Array();

	function queryHandler() {
		if (stdQuery.length == 0) {
			queryRunning = false;
			return true;
		}
		queryRunning = true;
		var value = stdQuery.shift();
		player.stdin.write(value);
		setTimeout(function () {
			 queryHandler()
		 }, 20);
	}

	function writeStdin (value) {

		if (open) {
			stdQuery.push(value);
			if (queryRunning == false ) {
				queryHandler();
			}
		} else {
			throw new Error('Player is closed.');
		}

	}

	// ----- Setup ----- //

	if (source) {
		player = spawnPlayer(source, output, loop, initialVolume, showOsd);

	}

	// ----- Methods ----- //

	// Restarts omxplayer with a new source.
	omxplayer.newSource = (src, out, loop, initialVolume, showOsd) => {

		if (open) {

			player.on('close', () => {
				omxProcess["pid"] = false
				player = spawnPlayer(src, out, loop, initialVolume, showOsd);
			});
			player.removeListener('close', updateStatus);
			writeStdin('q');

		} else {

			player = spawnPlayer(src, out, loop, initialVolume, showOsd);

		}


	};

	omxplayer.play = () => { writeStdin('p'); };
	omxplayer.pause = () => { writeStdin('p'); };
	omxplayer.volUp = () => { writeStdin('+'); };
	omxplayer.volDown = () => { writeStdin('-'); };
	omxplayer.fastFwd = () => { writeStdin('>'); };
	omxplayer.rewind = () => { writeStdin('<'); };
	omxplayer.fwd30 =() => { writeStdin('\u001b[C'); };
	omxplayer.back30 = () => { writeStdin('\u001b[D'); };
	omxplayer.fwd600 = () => { writeStdin('\u001b[A'); };
	omxplayer.back600 = () => { writeStdin('\u001b[B'); };
	omxplayer.quit = () => { writeStdin('q'); };
	omxplayer.subtitles = () => { writeStdin('s'); };
	omxplayer.info = () => { writeStdin('z'); };
	omxplayer.incSpeed = () => { writeStdin('1'); };
	omxplayer.decSpeed = () => { writeStdin('2'); };
	omxplayer.prevChapter = () => { writeStdin('i'); };
	omxplayer.nextChapter = () => { writeStdin('o'); };
	omxplayer.prevAudio = () => { writeStdin('j'); };
	omxplayer.nextAudio = () => { writeStdin('k'); };
	omxplayer.prevSubtitle = () => { writeStdin('n'); };
	omxplayer.nextSubtitle = () => { writeStdin('m'); };
	omxplayer.decSubDelay = () => { writeStdin('d'); };
	omxplayer.incSubDelay = () => { writeStdin('f'); };

	Object.defineProperty(omxplayer, 'running', {
		get: () => { return open; }
	});

	// ----- Return Object ----- //

	emitPlaying();
	return omxplayer;

}


// ----- Module Exports ----- //

module.exports = Omx;
